### PR TITLE
enhancement - working implementation on Windows10 with Visual Studio (...and Cygwin also)

### PR DESCRIPTION
**Pull request by keldonin**
Hi danni, it's me again :wink: 

I my previous PR, I claimed that it was working on the Windows platform. I had to revise this assertion; it indeed compiles (on Cygwin), but it does not work properly.

After digging, it is possible to make it work with Cygwin 32 bits, by adding a `#pragma pack(1)` before including the PKCS#11 header files; the situation is different with Cygwin 64 bits, as GCC `sizeof(CK_ULONG)` on this platform is different than on the 32 bits version, resulting in discrepancy for every structure. Since most of the PKCS#11 DLLs are built from MSVC, the interest is limited.

Actually, the `#pragma pack()` directive is mandatory on Windows platform, according to the PKCS#11 documentation.

I decided therefore to add the code needed to compile with MSVC. It uses the native Windows calls. It took me several changes:
- I had to remove the defines you set in `setup.py` and create a new header file that contains the pragmas
- there is also a new `_mswin.pxd` file that contains the definitions for Windows.
- I also had to use a couple of `IF` Cython precompiler directives.

I have tested on the Windows platform against SoftHSMv2 for Windows, and it passes most of the tests (see result - two tests have failed because of an issue with `subprocess.Popen()` on Windows).

I hope you'll enjoy it. This PR together with the previous should solve issues #12 and #27, IMO.

Eric

```bash
C:\Users\User\python-pkcs11>python setup.py test
running test
running egg_info
writing python_pkcs11.egg-info\PKG-INFO
writing dependency_links to python_pkcs11.egg-info\dependency_links.txt
writing requirements to python_pkcs11.egg-info\requires.txt
writing top-level names to python_pkcs11.egg-info\top_level.txt
writing manifest file 'python_pkcs11.egg-info\SOURCES.txt'
running build_ext
skipping 'pkcs11\_pkcs11.c' Cython extension (up-to-date)
copying build\lib.win-amd64-3.7\pkcs11\_pkcs11.cp37-win_amd64.pyd -> pkcs11
test_encrypt (tests.test_aes.AESTests) ... ok
test_encrypt_big_string (tests.test_aes.AESTests) ... ok
test_encrypt_stream (tests.test_aes.AESTests) ... ok
test_encrypt_whacky_sizes (tests.test_aes.AESTests) ... ok
test_sign (tests.test_aes.AESTests) ... skipped 'Requires Mechanism.AES_MAC'
test_sign_stream (tests.test_aes.AESTests) ... skipped 'Requires Mechanism.AES_MAC'
test_wrap (tests.test_aes.AESTests) ... ok
test_encrypt_des2 (tests.test_des.DESTests) ... ok
test_encrypt_des3 (tests.test_des.DESTests) ... ok
test_generate_des2_key (tests.test_des.DESTests) ... ok
test_generate_des3_key (tests.test_des.DESTests) ... ok
test_derive_key (tests.test_dh.DHTests) ... ok
test_generate_params (tests.test_dh.DHTests) ... ok
test_load_params (tests.test_dh.DHTests) ... ok
test_digest (tests.test_digest.DigestTests) ... ok
test_digest_generator (tests.test_digest.DigestTests) ... ok
test_digest_key (tests.test_digest.DigestTests) ... ok
test_digest_key_data (tests.test_digest.DigestTests) ... ok
test_generate_keypair_and_sign (tests.test_dsa.DSATests) ... ok
test_generate_keypair_directly (tests.test_dsa.DSATests) ... ok
test_generate_params (tests.test_dsa.DSATests) ... ok
test_derive_key (tests.test_ecc.ECCTests) ... ok
test_import_key_named_curve (tests.test_ecc.ECCTests) ... skipped 'Requires Mechanism.ECDSA_SHA1'
test_import_key_pair (tests.test_ecc.ECCTests) ... ok
test_import_key_params (tests.test_ecc.ECCTests) ... ok
test_sign_ecdsa (tests.test_ecc.ECCTests) ... ok
test_close_iterators (tests.test_iterators.IteratorTests) ... expected failure
test_partial_decrypt (tests.test_iterators.IteratorTests) ... ok
test_ecdh (tests.test_public_key_external.ExternalPublicKeyTests) ... ok
test_ecdsa (tests.test_public_key_external.ExternalPublicKeyTests) ... skipped 'Requires Mechanism.ECDSA_SHA1'
test_rsa (tests.test_public_key_external.ExternalPublicKeyTests) ... ok
test_terrible_hybrid_file_encryption_app (tests.test_public_key_external.ExternalPublicKeyTests) ... ok
test_encrypt_oaep (tests.test_rsa.RSATests) ... ok
test_encrypt_too_much_data (tests.test_rsa.RSATests) ... ok
test_key_wrap (tests.test_rsa.RSATests) ... ok
test_sign_default (tests.test_rsa.RSATests) ... ok
test_sign_pkcs_v15 (tests.test_rsa.RSATests) ... ok
test_sign_pss (tests.test_rsa.RSATests) ... ok
test_sign_stream (tests.test_rsa.RSATests) ... ok
test_copy_object (tests.test_sessions.SessionTests) ... skipped 'SoftHSMv2 only'
test_create_object (tests.test_sessions.SessionTests) ... ok
test_destroy_object (tests.test_sessions.SessionTests) ... ok
test_generate_key (tests.test_sessions.SessionTests) ... ok
test_generate_keypair (tests.test_sessions.SessionTests) ... ok
test_generate_random (tests.test_sessions.SessionTests) ... ok
test_get_key (tests.test_sessions.SessionTests) ... ok
test_get_key_not_found (tests.test_sessions.SessionTests) ... ok
test_get_key_vague (tests.test_sessions.SessionTests) ... ok
test_get_objects (tests.test_sessions.SessionTests) ... ok
test_open_session (tests.test_sessions.SessionTests) ... ok
test_open_session_and_login_so (tests.test_sessions.SessionTests) ... skipped 'SoftHSMv2 only'
test_open_session_and_login_user (tests.test_sessions.SessionTests) ... ok
test_seed_random (tests.test_sessions.SessionTests) ... ok
test_double_initialise (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_double_initialise_different_libs (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_mechanism_info (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_mechanism_info_ec (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_mechanisms (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_slots (tests.test_slots_and_tokens.SlotsAndTokensTests) ... skipped 'SoftHSMv2 only'
test_get_token (tests.test_slots_and_tokens.SlotsAndTokensTests) ... skipped 'SoftHSMv2 only'
test_get_tokens (tests.test_slots_and_tokens.SlotsAndTokensTests) ... skipped 'SoftHSMv2 only'
test_concurrency (tests.test_threading.ThreadingTests) ... ok
test_import_ca_certificate (tests.test_x509.X509Tests) ... ok
test_import_ca_certificate_easy (tests.test_x509.X509Tests) ... ok
test_self_sign_certificate (tests.test_x509.X509Tests) ... C:\Users\User\AppData\Local\Programs\Python\Python37\lib\subprocess.py:664: ResourceWarning: unclosed file <_io.BufferedReader name=5>
  _cleanup()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
ERROR
test_sign_csr (tests.test_x509.X509Tests) ... ERROR
test_verify_certificate_dsa (tests.test_x509.X509Tests) ... ok
test_verify_certificate_ecdsa (tests.test_x509.X509Tests) ... skipped 'Requires Mechanism.ECDSA_SHA1'
test_verify_certificate_rsa (tests.test_x509.X509Tests) ... ok

======================================================================
ERROR: test_self_sign_certificate (tests.test_x509.X509Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\user\python-pkcs11\tests\__init__.py", line 73, in wrapper
    return func(self, *args, **kwargs)
  File "c:\users\user\python-pkcs11\tests\test_x509.py", line 217, in test_self_sign_certificate
    stdout=subprocess.DEVNULL) as proc:
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\subprocess.py", line 1178, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] Le fichier spécifié est introuvable

======================================================================
ERROR: test_sign_csr (tests.test_x509.X509Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\user\python-pkcs11\tests\__init__.py", line 73, in wrapper
    return func(self, *args, **kwargs)
  File "c:\users\user\python-pkcs11\tests\test_x509.py", line 262, in test_sign_csr
    stdout=subprocess.DEVNULL) as proc:
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Users\User\AppData\Local\Programs\Python\Python37\lib\subprocess.py", line 1178, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] Le fichier spécifié est introuvable

----------------------------------------------------------------------
Ran 69 tests in 1.625s

FAILED (errors=2, skipped=10, expected failures=1)
Test failed: <unittest.runner.TextTestResult run=69 errors=2 failures=0>
error: Test failed: <unittest.runner.TextTestResult run=69 errors=2 failures=0>
```

### CODE REVIEW

File: pkcs11/_pkcs11.pyx [hunk 1]
```diff
@@ -11,11 +11,14 @@ library loaded.
...
 
-from posix cimport dlfcn
-
 from cython.view cimport array
 from cpython.mem cimport PyMem_Malloc, PyMem_Free
 
+IF UNAME_SYSNAME == "Windows":
+    from . cimport _mswin as mswin
+ELSE:
+    from posix cimport dlfcn
```
Comment (danni) [inline pkcs11/_pkcs11.pyx]: Surprised no one has wrapped this actually.

File: pkcs11/_pkcs11.pyx [hunk 2]
```diff
@@ -1187,16 +1194,50 @@ cdef class lib:
...
+
+        IF UNAME_SYSNAME == "Windows":
+            if self._handle != NULL:
+                mswin.FreeLibrary(self._handle)
+        ELSE:
+            if self._handle != NULL:
+                dlfcn.dlclose(self._handle)
+
+
+    IF UNAME_SYSNAME == "Windows":
```
Comment (danni) [inline pkcs11/_pkcs11.pyx]: Should this go in your `mswin.pxd`?

File: setup.py [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
 """
 setup.py
 """
 
 from setuptools import setup, find_packages
 from setuptools.extension import Extension
+from platform import system
+
+# if compiling using MSVC, we need to add user32 library
+libraries = ('user32',) if system() == 'Windows' else ()
```
Comment (danni) [inline setup.py]: Avoid ternary operators.

File: extern/cryptoki.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(__CYGWIN64__)
+#pragma warning "Cygwin 64 bits build will only work with Cygwin64-compiled PKCS#11 modules"
+#endif
+
+#define CK_PTR            *
```
Comment (danni) [inline extern/cryptoki.h]: So I understood these defines were different on the different platforms. Is that not the case?

**Comment by danni** [comment]
Are the tests failing because it can't find `openssl`? We can probably add a decorator to `@skip` if `openssl` is not found.

It looks like the the `Is` class used for decorating some tests is broken on Windows. There's a few tests being skipped with `skipped 'SoftHSMv2 only'`. I'm guessing this is because it hard-codes the library name, which is different on Windows.

File: pkcs11/_pkcs11.pyx (hunk 1 above)
Comment (keldonin) [reply pkcs11/_pkcs11.pyx]: Not sure to get your point here (sorry, I'm not a native speaker). Could you clarify please?

File: pkcs11/_pkcs11.pyx (hunk 2 above)
Comment (keldonin) [reply pkcs11/_pkcs11.pyx]: I am assuming you mean the `cdef _winerrormsg()` code section surrounded with the `IF` statement.
I have a personal preference to put only headers declaration into `.pxd` files; Cython however allows code sections into these files. Fair enough, I'll move this as per request.

File: setup.py (hunk 3 above)
Comment (keldonin) [reply setup.py]: Fair enough.

File: extern/cryptoki.h (hunk 4 above)
Comment (keldonin) [reply extern/cryptoki.h]: Short answer: not really...

Long one - take a deep breath:
PKCS#11 was originally designed with Windows platform in mind. ( and, by the time, when 16 bits binaries was still the norm, presumably). You had the ability to define many kinds of pointers ( e.g. far pointers vs short pointers). Hence the assumption these defines could change.
If we consider Today's platforms share, you mainly have Windows and unix _ish_ ones. Incidentally, the same defines apply to all platform, for all the defines.

As an evidence, feel free to checkout this code on Github: https://github.com/Mastercard/pkcs11-tools/blob/master/include/cryptoki/cryptoki.h , from my project (making free advertisement, I recognize :wink: ) that compile nicely on Windows(cross-compiled), Linux, Freebsd, AIX, Solaris, MacOS,... All sharing the same header file.

Actually, PKCS#11 screwed up with one thing: defining `CK_ULONG` as `unsigned int long`, which yields a 64 bits wide integer on unix _ish_ patforms. This is not an issue when the PKCS#11 drivers are all compiled through the same compiler that we use; it leads to discrepancies when e.g. you start using Cygwin64 - which implements gcc 64 bits like on a unix platform, and try to interface that code with MSVC-compiled PKCS#11 DLLs, where `sizeof(unsigned long int)` is always 32 bits, irrespective of the processor bit width.

Hope this answers your question.

**Comment by keldonin** [comment]
I am currently abroad, and don't have the ability to progress on this, this week. Thanks for the questions/comments, I'll work this out when I'm back and push an update to the PR.

File: pkcs11/_pkcs11.pyx (hunk 1 above)
Comment (danni) [reply pkcs11/_pkcs11.pyx]: I'm surprised there's not a cross-platform library open for Cython.

File: extern/cryptoki.h (hunk 4 above)
Comment (danni) [reply extern/cryptoki.h]: Thanks for that detailed answer.

File: pkcs11/_pkcs11.pyx (hunk 1 above)
Comment (keldonin) [reply pkcs11/_pkcs11.pyx]: Oh OK. That would be great indeed!  I must admit I didn't look for it. If not existing, this is perhaps opportunity for a new project...

**Comment by keldonin** [comment]
There you go. For testing, I took the freedom to enhance a few side items:
- environment variables for SO PIN (instead of hard-coded value), and defaulting to user PIN
- detection of a library is now multi-platform (Windows, MacOS, linux)
- detection for openssl binary, and skipping these tests if not found
- path to openssl can be specified as an environment variable, if not in the path

Here is the result: (using softHSM2)

```
C:\Users\User\python-pkcs11>set
...
OPENSSL_PATH=C:\OpenSSL-Win64\bin
...
PKCS11_MODULE=C:\SoftHSM2\lib\softhsm2-x64.dll
PKCS11_TOKEN_LABEL=token-1
PKCS11_TOKEN_PIN=changeit
PKCS11_TOKEN_SO_PIN=changeit
...

C:\Users\User\python-pkcs11>python setup.py test
running test
running egg_info
writing python_pkcs11.egg-info\PKG-INFO
writing dependency_links to python_pkcs11.egg-info\dependency_links.txt
writing requirements to python_pkcs11.egg-info\requires.txt
writing top-level names to python_pkcs11.egg-info\top_level.txt
writing manifest file 'python_pkcs11.egg-info\SOURCES.txt'
running build_ext
cythoning pkcs11/_pkcs11.pyx to pkcs11\_pkcs11.c
building 'pkcs11._pkcs11' extension
creating build
creating build\temp.win-amd64-3.7
creating build\temp.win-amd64-3.7\Release
creating build\temp.win-amd64-3.7\Release\pkcs11
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -IC:\Users\User\AppData\Local\Programs\Python\Python37\include -IC:\Users\User\AppData\Local\Programs\Python\Python37\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\cppwinrt" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\cppwinrt" /Tcpkcs11\_pkcs11.c /Fobuild\temp.win-amd64-3.7\Release\pkcs11\_pkcs11.obj
_pkcs11.c
creating C:\Users\User\python-pkcs11\build\lib.win-amd64-3.7
creating C:\Users\User\python-pkcs11\build\lib.win-amd64-3.7\pkcs11
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:C:\Users\User\AppData\Local\Programs\Python\Python37\libs /LIBPATH:C:\Users\User\AppData\Local\Programs\Python\Python37\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\um\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\um\x64" user32.lib /EXPORT:PyInit__pkcs11 build\temp.win-amd64-3.7\Release\pkcs11\_pkcs11.obj /OUT:build\lib.win-amd64-3.7\pkcs11\_pkcs11.cp37-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.7\Release\pkcs11\_pkcs11.cp37-win_amd64.lib
   Creating library build\temp.win-amd64-3.7\Release\pkcs11\_pkcs11.cp37-win_amd64.lib and object build\temp.win-amd64-3.7\Release\pkcs11\_pkcs11.cp37-win_amd64.exp
Generating code
Finished generating code
copying build\lib.win-amd64-3.7\pkcs11\_pkcs11.cp37-win_amd64.pyd -> pkcs11
test_encrypt (tests.test_aes.AESTests) ... ok
test_encrypt_big_string (tests.test_aes.AESTests) ... ok
test_encrypt_stream (tests.test_aes.AESTests) ... ok
test_encrypt_whacky_sizes (tests.test_aes.AESTests) ... ok
test_sign (tests.test_aes.AESTests) ... skipped 'Requires Mechanism.AES_MAC'
test_sign_stream (tests.test_aes.AESTests) ... skipped 'Requires Mechanism.AES_MAC'
test_wrap (tests.test_aes.AESTests) ... ok
test_encrypt_des2 (tests.test_des.DESTests) ... ok
test_encrypt_des3 (tests.test_des.DESTests) ... ok
test_generate_des2_key (tests.test_des.DESTests) ... ok
test_generate_des3_key (tests.test_des.DESTests) ... ok
test_derive_key (tests.test_dh.DHTests) ... ok
test_generate_params (tests.test_dh.DHTests) ... ok
test_load_params (tests.test_dh.DHTests) ... ok
test_digest (tests.test_digest.DigestTests) ... ok
test_digest_generator (tests.test_digest.DigestTests) ... ok
test_digest_key (tests.test_digest.DigestTests) ... ok
test_digest_key_data (tests.test_digest.DigestTests) ... ok
test_generate_keypair_and_sign (tests.test_dsa.DSATests) ... ok
test_generate_keypair_directly (tests.test_dsa.DSATests) ... ok
test_generate_params (tests.test_dsa.DSATests) ... ok
test_derive_key (tests.test_ecc.ECCTests) ... ok
test_import_key_named_curve (tests.test_ecc.ECCTests) ... skipped 'Requires Mechanism.ECDSA_SHA1'
test_import_key_pair (tests.test_ecc.ECCTests) ... ok
test_import_key_params (tests.test_ecc.ECCTests) ... ok
test_sign_ecdsa (tests.test_ecc.ECCTests) ... ok
test_close_iterators (tests.test_iterators.IteratorTests) ... expected failure
test_partial_decrypt (tests.test_iterators.IteratorTests) ... ok
test_ecdh (tests.test_public_key_external.ExternalPublicKeyTests) ... ok
test_ecdsa (tests.test_public_key_external.ExternalPublicKeyTests) ... skipped 'Requires Mechanism.ECDSA_SHA1'
test_rsa (tests.test_public_key_external.ExternalPublicKeyTests) ... ok
test_terrible_hybrid_file_encryption_app (tests.test_public_key_external.ExternalPublicKeyTests) ... ok
test_encrypt_oaep (tests.test_rsa.RSATests) ... ok
test_encrypt_too_much_data (tests.test_rsa.RSATests) ... ok
test_key_wrap (tests.test_rsa.RSATests) ... ok
test_sign_default (tests.test_rsa.RSATests) ... ok
test_sign_pkcs_v15 (tests.test_rsa.RSATests) ... ok
test_sign_pss (tests.test_rsa.RSATests) ... ok
test_sign_stream (tests.test_rsa.RSATests) ... ok
test_copy_object (tests.test_sessions.SessionTests) ... ok
test_create_object (tests.test_sessions.SessionTests) ... ok
test_destroy_object (tests.test_sessions.SessionTests) ... ok
test_generate_key (tests.test_sessions.SessionTests) ... ok
test_generate_keypair (tests.test_sessions.SessionTests) ... ok
test_generate_random (tests.test_sessions.SessionTests) ... ok
test_get_key (tests.test_sessions.SessionTests) ... ok
test_get_key_not_found (tests.test_sessions.SessionTests) ... ok
test_get_key_vague (tests.test_sessions.SessionTests) ... ok
test_get_objects (tests.test_sessions.SessionTests) ... ok
test_open_session (tests.test_sessions.SessionTests) ... ok
test_open_session_and_login_so (tests.test_sessions.SessionTests) ... ok
test_open_session_and_login_user (tests.test_sessions.SessionTests) ... ok
test_seed_random (tests.test_sessions.SessionTests) ... ok
test_double_initialise (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_double_initialise_different_libs (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_mechanism_info (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_mechanism_info_ec (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_mechanisms (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_slots (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_token (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_get_tokens (tests.test_slots_and_tokens.SlotsAndTokensTests) ... ok
test_concurrency (tests.test_threading.ThreadingTests) ... ok
test_import_ca_certificate (tests.test_x509.X509Tests) ... ok
test_import_ca_certificate_easy (tests.test_x509.X509Tests) ... ok
test_self_sign_certificate (tests.test_x509.X509Tests) ... ok
test_sign_csr (tests.test_x509.X509Tests) ... verify OK
ok
test_verify_certificate_dsa (tests.test_x509.X509Tests) ... ok
test_verify_certificate_ecdsa (tests.test_x509.X509Tests) ... skipped 'Requires Mechanism.ECDSA_SHA1'
test_verify_certificate_rsa (tests.test_x509.X509Tests) ... ok

----------------------------------------------------------------------
Ran 69 tests in 1.702s

OK (skipped=5, expected failures=1)
```

**Comment by danni** [comment]
Wait a moment. I'm not sure why Travis has stopped reporting, but there's something to fix up: https://travis-ci.org/danni/python-pkcs11/jobs/493412106

**Comment by keldonin** [comment]
😅 Here is the reason:
you hard-coded the SO PIN to `'5678'`, in the current implementation, see https://github.com/danni/python-pkcs11/blob/82cf55683bedc7a662de7351d5db9ff07bb6248d/tests/test_sessions.py#L26 
The PR no longer assumes this value and either expects that you set it through a new environment variable `PKCS11_TOKEN_SO_PIN`, or assumes it be the same as for `PKCS11_TOKEN_PIN`; see `tests/__init__.py` from PR, it's documented in the module docstring.

Hence Travis legitimately complaining:

- You set the SO PIN here: https://travis-ci.org/danni/python-pkcs11/jobs/493412106#L2242
- Test code from this PR warns about `PKCS11_TOKEN_SO_PIN` unset, and assumes the value is the same as for `PKCS11_TOKEN_PIN`: https://travis-ci.org/danni/python-pkcs11/jobs/493412106#L2270
- The test `test_open_session_and_login_so` miserably fails with a `PinIncorrect` exception: https://travis-ci.org/danni/python-pkcs11/jobs/493412106#L2284

Too bad that Travis triggers only upon changes approval and not upon PR submission - I would have spotted it. (Can Github be configured that way?)

I have adapted Travis config file to fix the issue, in my last commit. Travis seems happy now.

**Comment by danni** [comment]
I'm sure it used to, maybe it stopped because people were executing random denials of service.

**Comment by danni** [comment]
Thanks again for this. I'll try and do a release on the weekend.

**Comment by avoidik** [comment]
@danni kindly reminder ;)

**Comment by danni** [comment]
Thanks @avoidik. Just trying to figure out Windows CI.

**Pull request merged**